include/ppx/metrics.h

```diff
@@ -0,0 +1,198 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ppx_metrics_h
+#define ppx_metrics_h
+
+#include "ppx/config.h"
+
+#include <limits>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace ppx {
+namespace metrics {
+
+#define METRICS_NO_COPY(type)   \
+    type(type&)       = delete; \
+    type(const type&) = delete; \
+    type& operator=(const type&) = delete;
+
+////////////////////////////////////////////////////////////////////////////////
+
+enum class MetricInterpretation
+{
+    NONE,
+    HIGHER_IS_BETTER,
+    LOWER_IS_BETTER,
+};
+
+struct Range
+{
+    double lowerBound = std::numeric_limits<double>::min();
+    double upperBound = std::numeric_limits<double>::max();
+};
+
+struct MetricMetadata
+{
+    std::string          name;
+    std::string          unit;
+    MetricInterpretation interpretation = MetricInterpretation::NONE;
+    Range                expectedRange;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+enum class MetricType
+{
+    GAUGE,
+    COUNTER,
+};
+
+class Metric
+{
+    friend class Run;
+
+public:
+    const MetricMetadata& GetMetadata() const;
+    MetricType            GetType() const;
+
+protected:
+    Metric(const MetricMetadata& metadata, MetricType type);
+    virtual ~Metric();
+
+private:
+    MetricMetadata mMetadata;
+    MetricType     mType;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+struct GaugeStatistics
+{
+    double min;
+    double max;
+    double average;
+    double median;
+    double standardDeviation;
+    double timeRatio;
+    double percentile90;
+    double percentile95;
+    double percentile99;
+};
+
+class MetricGauge final : public Metric
+{
+public:
+    MetricGauge(const MetricMetadata& metadata);
+
+    void   RecordEntry(double seconds, double value);
+    size_t GetEntriesCount() const;
+    void   GetEntry(size_t index, double& seconds, double& value) const;
```

Comment:
Fixed!

Reply:
Oops, forgot those. I turned them into pointers. As for the naming, let me do it in another CL. I don't think I used any "p" prefix anywhere.
Is the naming a BW convention? I'd rather not mingle the type with the name of variables precisely because when you refactor, you have to rename variables.
I will comply to whatever BW rules there is though.

Reply:
It's a bigwheels convention (mostly observed across the project as it's not enforced by tools). It's to follow Microsoft's convention:
 - pMyPointer
 - ppMyPointerOfPointer
 - pppMyPointerOfPointerOfPointer

Reply:
OK, let me fix all these in a separate, dedicated, CL then.